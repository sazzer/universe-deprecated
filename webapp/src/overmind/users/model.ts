/** The state of the user record */
export type UserState = "LOADING" | "SAVING" | "UNKNOWN" | "LOADED";

/** The shape of a single user */
export interface User {
  /** The unqiue ID of the user */
  id: string;
  /** The unique username of the user */
  username: string;
  /** The display name of the user */
  displayName: string;
  /** The email address of the user, if we know what it is */
  email: string | null;
}

/** Representation of a user as known by the state */
export interface StoredUser {
  /** The user record, if we've currently got one */
  user: User | undefined;

  /** The current state of the user */
  state: UserState;
}

/** Error when loading a user that doesn't exist */
export class UnknownUserError extends Error {}
