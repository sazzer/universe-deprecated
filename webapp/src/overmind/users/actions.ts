import { Action } from "overmind";
import { User } from "./model";

/**
 * Overmind action for storing the details of a user
 */
export const storeUser: Action<User> = ({ state }, details: User) => {
  state.users.users[details.id] = {
    id: details.id,
    username: details.username,
    displayName: details.displayName,
    email: details.email
  };
  delete state.users.userStates[details.id];
};

export const fetchUser: Action<string, Promise<void>> = async (
  { state, actions, effects },
  userId: string
) => {
  state.users.userStates[userId] = "LOADING";
  const user = await effects.users.api.loadUser(userId);
  actions.users.storeUser(user);
};
