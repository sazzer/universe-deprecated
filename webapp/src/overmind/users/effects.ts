import { request, ProblemResponse } from "../../api";
import { User, UnknownUserError } from "./model";

/**
 * Retrieve a user from the API
 * @param id The ID of the user to retrieve
 * @return The details of the user
 */
async function loadUser(id: string): Promise<User> {
  try {
    const user = await request<User>({
      url: "/users/{id}",
      urlParams: {
        id
      },
      method: "GET"
    });

    return user.data;
  } catch (e) {
    throw new UnknownUserError();
  }
}

/** The API for working with users*/
export const api = {
  loadUser
};
