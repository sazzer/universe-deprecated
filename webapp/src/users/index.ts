export * from "./model";
export * from "./userContext";
export * from "./api";
