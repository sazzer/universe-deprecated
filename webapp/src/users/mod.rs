mod model;
mod service;

pub use model::*;
