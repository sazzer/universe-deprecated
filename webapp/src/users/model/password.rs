use bcrypt::{hash, verify, BcryptError, DEFAULT_COST};
use bytes::BytesMut;
use log::warn;
use postgres::types::{accepts, to_sql_checked, FromSql, IsNull, ToSql, Type};
use serde::{Serialize, Serializer};
use std::error::Error;

/// Representation of a password that has been securely hashed
#[derive(Debug, PartialEq, Clone)]
pub struct Password(String);

/// Enumeration of errors that can occur when hashing a password
#[derive(Debug, PartialEq)]
pub enum PasswordHashError {
    HashError,
}

impl From<BcryptError> for PasswordHashError {
    fn from(_: BcryptError) -> Self {
        PasswordHashError::HashError
    }
}

impl Password {
    /// Construct a new Password instance from an already hashed string.
    ///
    /// # Arguments
    /// * `hash` The hash to wrap
    ///
    /// # Returns
    /// The `Password` to wrap the hash
    pub fn from_hash<S>(hash: S) -> Password
    where
        S: Into<String>,
    {
        Password(hash.into())
    }

    /// Generate a new Password instance from an unhashed password
    ///
    /// # Arguments
    /// * `plaintext` The password to hash
    ///
    /// # Returns
    /// The hashed `Password` representing the provided plaintext
    pub fn from_plaintext<S>(plaintext: S) -> Result<Password, PasswordHashError>
    where
        S: Into<String>,
    {
        let hashed = hash(plaintext.into(), DEFAULT_COST)?;
        Ok(Password(hashed))
    }

    /// Verify if our hashed password is consistent with the provided plaintext.
    ///
    /// # Arguments
    /// * `plaintext` The plaintext to compare against
    ///
    /// # returns
    /// True if the provided plaintext is consistent. False if not.
    pub fn verify<S>(&self, plaintext: S) -> bool
    where
        S: Into<String>,
    {
        match verify(plaintext.into(), &self.0) {
            Ok(v) => v,
            Err(e) => {
                warn!("Error verifying password: {}", e);
                false
            }
        }
    }
}

/// Allow us to serialize `Password` objects into Serde structures.
///
/// This always outputs `null` regardless of the password, so that we can never accidentally leak
/// hashed passwords out to clients.
impl Serialize for Password {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_none()
    }
}

/// Allow us to pass `Password` objects to Postgres as part of executing a database query.
///
/// The implementation of this trait allows objects of this type to be used directly as database
/// binds without ever needing to extract the string from inside it.
impl ToSql for Password {
    fn to_sql(&self, t: &Type, w: &mut BytesMut) -> Result<IsNull, Box<dyn Error + Sync + Send>> {
        self.0.to_sql(t, w)
    }

    accepts!(VARCHAR, TEXT);
    to_sql_checked!();
}

/// Allow us to retrieve `Password` objects from Postgres as part of executing a database query.
///
/// The implementation of this trait allows objects of this type to be read directly as database
/// outputs without needing to construct it explicitly. Instead the Postgres crate will do so for us.
impl<'a> FromSql<'a> for Password {
    fn from_sql(t: &Type, raw: &'a [u8]) -> Result<Self, Box<dyn Error + Sync + Send>> {
        String::from_sql(t, raw).map(Password)
    }
    accepts!(VARCHAR, TEXT);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::database::test::TestDatabase;
    use postgres::Error;
    use spectral::prelude::*;

    #[test]
    fn test_hashing_simple_password() {
        let password = Password::from_plaintext("password");

        assert_that(&password)
            .is_ok()
            .is_not_equal_to(Password::from_hash("password"));
    }

    #[test]
    fn test_hashing_invalid_password() {
        let plaintext = std::str::from_utf8(&[65u8, 66u8, 0u8, 67u8, 68u8]).unwrap();
        let password = Password::from_plaintext(plaintext);

        assert_that(&password)
            .is_err()
            .is_equal_to(PasswordHashError::HashError);
    }

    #[test]
    fn test_verify_valid_password() {
        let password = Password::from_plaintext("password").unwrap();
        let result = password.verify("password");
        assert_that(&result).is_equal_to(true);
    }

    #[test]
    fn test_verify_wrong_password() {
        let password = Password::from_plaintext("password").unwrap();
        let result = password.verify("wrong");
        assert_that(&result).is_equal_to(false);
    }

    #[test]
    fn test_verify_wrong_case() {
        let password = Password::from_plaintext("password").unwrap();
        let result = password.verify("Password");
        assert_that(&result).is_equal_to(false);
    }

    #[test]
    fn test_verify_invalid_hash() {
        let password = Password::from_hash("password_hash");
        let result = password.verify("password_hash");
        assert_that(&result).is_equal_to(false);
    }

    #[test]
    fn test_serde_serialize() {
        let password = Password::from_hash("password_hash");

        let serialized = serde_json::to_value(password);
        assert_that(&serialized)
            .is_ok()
            .is_equal_to(serde_json::Value::Null);
    }

    #[test]
    fn test_postgres_to_sql() {
        let database = TestDatabase::new();

        let password = Password::from_hash("password_hash");
        let result = database.client().query("SELECT $1", &[&password]);

        let rows = result.unwrap();
        assert_that(&rows.len()).is_equal_to(1);

        let row = rows.get(0).unwrap();
        assert_that(&row.len()).is_equal_to(1);

        let output_value: &str = rows.get(0).unwrap().get(0);
        assert_that(&output_value).is_equal_to("password_hash");
    }

    #[test]
    fn test_postgres_from_sql_valid_type() {
        let database = TestDatabase::new();

        let result = database.client().query("SELECT $1", &[&"password_hash"]);

        let rows = result.unwrap();
        assert_that(&rows.len()).is_equal_to(1);

        let row = rows.get(0).unwrap();
        assert_that(&row.len()).is_equal_to(1);

        let output_value: Password = rows.get(0).unwrap().get(0);
        assert_that(&output_value).is_equal_to(Password::from_hash("password_hash"));
    }

    #[test]
    fn test_postgres_from_sql_invalid_type() {
        let database = TestDatabase::new();

        let result = database.client().query("SELECT $1", &[&1]);

        let err: Error = result.err().unwrap();
        let err_msg = format!("{}", err);
        assert_that(&err_msg)
                    .is_equal_to("error serializing parameter 0: cannot convert between the Rust type `i32` and the Postgres type `text`".to_owned());
    }
}
