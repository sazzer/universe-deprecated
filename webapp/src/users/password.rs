/// Representation of a securely hashed password that we can compare to other passwords to see if
/// they match
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Password(String);

impl Password {
    /// Construct a new `Password` from the already hashed string - i.e. when loading from the data
    /// store.
    ///
    /// # Arguments
    /// * `hash` The hash that we are wrapping in a `Password`
    ///
    /// # Returns
    /// The `Password` wrapper around the hash
    pub fn from_hash<S: Into<String>>(hash: S) -> Self {
        Password(hash.into())
    }
}
