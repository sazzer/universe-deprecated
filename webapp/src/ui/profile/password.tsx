import * as yup from "yup";

import { ErrorMessage, FieldValues, useForm } from "react-hook-form";
import React, { useState } from "react";
import { changePassword, useUser } from "../../users";

import { SubmitButton } from "../components/form/buttons";
import { UnexpectedError } from "../components/form/error";
import { ValidationErrors } from "../../api";
import debug from "debug";
import { useTranslation } from "react-i18next";

/** The logger to use */
const LOG = debug("universe:ui:profile:changePassword");

/**
 * The actual Change Password form to work with
 */
export const ChangePasswordArea: React.FC = () => {
  const { t } = useTranslation();
  const [loading, setLoading] = useState(false);
  const [globalError, setGlobalError] = useState<string | undefined>(undefined);
  const { user } = useUser();

  const { register, errors, handleSubmit, setError } = useForm({
    validationSchema: yup.object().shape({
      password: yup
        .string()
        .required(
          t(
            "profile.password.password.errors.tag:universe,2020:validation-errors/missing"
          )
        ),
      password2: yup
        .string()
        .required(
          t(
            "profile.password.password2.errors.tag:universe,2020:validation-errors/missing"
          )
        )
        .when(["password"], (password: string, schema: any) => {
          return schema.oneOf(
            [password],
            t(
              "profile.password.password2.errors.tag:universe,2020:validation-errors/password/invalid-password"
            )
          );
        })
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      password: "",
      password2: ""
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    LOG("Submitting form: %o", data);
    setGlobalError(undefined);
    setLoading(true);

    try {
      const saved = await changePassword(user?.id || "", data.password);
    } catch (e) {
      if (e instanceof ValidationErrors) {
        e.errors.forEach(error => {
          const message = t(
            `profile.profile.${error.field}.errors.${error.type}`
          );
          setError(error.field, error.type, message);
        });
      } else {
        setGlobalError(e.toString());
      }
    }
    setLoading(false);
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmitHandler)}
      data-test="changePasswordForm"
    >
      <div className="form-group" data-test="password">
        <label htmlFor="password">{t("profile.password.password.label")}</label>
        <input
          type="password"
          className={
            errors.password ? "form-control is-invalid" : "form-control"
          }
          id="password"
          name="password"
          readOnly={loading}
          autoFocus
          ref={register}
        />
        <ErrorMessage
          errors={errors}
          name="password"
          as={<div className="invalid-feedback" />}
        />
      </div>
      <div className="form-group" data-test="password2">
        <label htmlFor="password2">
          {t("profile.password.password2.label")}
        </label>
        <input
          type="password"
          className={
            errors.password2 ? "form-control is-invalid" : "form-control"
          }
          id="password2"
          name="password2"
          readOnly={loading}
          ref={register}
        />
        <ErrorMessage
          errors={errors}
          name="password2"
          as={<div className="invalid-feedback" />}
        />
      </div>
      <div className="form-group">
        <SubmitButton loading={loading}>
          {t("profile.password.submit")}
        </SubmitButton>
      </div>
      {globalError && <UnexpectedError message={globalError} />}
    </form>
  );
};
