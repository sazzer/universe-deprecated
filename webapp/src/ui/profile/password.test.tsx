import { changePassword, useUser } from "../../users";
import { fireEvent, render, wait } from "@testing-library/react";

import { ChangePasswordArea } from "./password";
import React from "react";

jest.mock("../../users");

const useUserMock = useUser as jest.Mock;
const changePasswordMock = changePassword as jest.Mock;

afterEach(() => {
  jest.resetAllMocks();
});

test("Rendering after user loaded", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  const { container } = render(<ChangePasswordArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  expect(container).toMatchSnapshot();

  expect(storeUser).toBeCalledTimes(0);
  expect(changePasswordMock).toBeCalledTimes(0);
});

test("Saving blank fields", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  const { container, getByText, getByLabelText } = render(
    <ChangePasswordArea />
  );

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: "" }
    });
    fireEvent.change(getByLabelText("Repeat Password"), {
      target: { value: "" }
    });
    fireEvent.click(getByText("Change Password", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(0);
  expect(changePasswordMock).toBeCalledTimes(0);
});

test("Saving mismatched passwords", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  const { container, getByText, getByLabelText } = render(
    <ChangePasswordArea />
  );

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: "Pa55word" }
    });
    fireEvent.change(getByLabelText("Repeat Password"), {
      target: { value: "password" }
    });
    fireEvent.click(getByText("Change Password", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(0);
  expect(changePasswordMock).toBeCalledTimes(0);
});
test("Saving successfully", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  changePasswordMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "newuser@example.com",
    username: "testuser",
    displayName: "New User"
  });

  const { container, getByText, getByLabelText } = render(
    <ChangePasswordArea />
  );

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: "Pa55word" }
    });
    fireEvent.change(getByLabelText("Repeat Password"), {
      target: { value: "Pa55word" }
    });
    fireEvent.click(getByText("Change Password", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(0);
  expect(changePasswordMock).toBeCalledTimes(1);
  expect(changePasswordMock).toBeCalledWith(
    "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    "Pa55word"
  );
});

test("Saving with an unexpected error", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  changePasswordMock.mockRejectedValueOnce(new Error("Network error"));

  const { container, getByText, getByLabelText } = render(
    <ChangePasswordArea />
  );

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("New Password"), {
      target: { value: "Pa55word" }
    });
    fireEvent.change(getByLabelText("Repeat Password"), {
      target: { value: "Pa55word" }
    });
    fireEvent.click(getByText("Change Password", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(0);
  expect(changePasswordMock).toBeCalledTimes(1);
  expect(changePasswordMock).toBeCalledWith(
    "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    "Pa55word"
  );
});
