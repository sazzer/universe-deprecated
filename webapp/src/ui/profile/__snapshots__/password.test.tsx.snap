// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering after user loaded 1`] = `
<div>
  <form
    data-test="changePasswordForm"
  >
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        New Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving blank fields 1`] = `
<div>
  <form
    data-test="changePasswordForm"
  >
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        New Password
      </label>
      <input
        class="form-control is-invalid"
        id="password"
        name="password"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a password
      </div>
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control is-invalid"
        id="password2"
        name="password2"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Please re-enter the password
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving mismatched passwords 1`] = `
<div>
  <form
    data-test="changePasswordForm"
  >
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        New Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control is-invalid"
        id="password2"
        name="password2"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Passwords do not match
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving successfully 1`] = `
<div>
  <form
    data-test="changePasswordForm"
  >
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        New Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-success"
        role="alert"
      >
        Password changed successfully
      </div>
    </div>
  </form>
</div>
`;

exports[`Saving with an unexpected error 1`] = `
<div>
  <form
    data-test="changePasswordForm"
  >
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        New Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
      data-test="password2"
    >
      <label
        for="password2"
      >
        Repeat Password
      </label>
      <input
        class="form-control"
        id="password2"
        name="password2"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Change Password
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-danger"
        role="alert"
      >
        An unexpected error occurred: Error: Network error
      </div>
    </div>
  </form>
</div>
`;
